tgui/packages/tgui/interfaces/TechFab.js

```diff
@@ -0,0 +1,364 @@
+  );
+};
+
+const TechFabHeader = (props, context) => {
+  const { act, data } = useBackend(context);
+  const {
+    materials = {},
+    materials_label = "0/unlimited",
+    reagents = {},
+    reagents_label = "",
+  } = data;
+
+  return (
+    <Stack.Item>
+      <Section>
+        <Collapsible
+          title={"Materials ("+materials_label+")"}
+          disabled={materials === null}>
+          <Flex wrap="wrap" align="baseline">
+            {
+              materials && Object.keys(materials).map(id => {
+                const material = materials[id];
+
+                return (
+                  <Material key={id} material={material} />
+                );
+              })
+            }
+          </Flex>
+        </Collapsible>
+        <Collapsible
+          title={"Reagents ("+reagents_label+")"}
+          disabled={materials === null}>
+          <Flex wrap="wrap" align="baseline">
+            <Flex.Item width="100%">
+              <Button
+                content="Purge all" 
+                onClick={() => act("disposeall", {})}
+              />
+            </Flex.Item>
+            {
+              reagents && Object.keys(reagents).map(id => {
+                const reagent = reagents[id];
+
+                return (
+                  <Reagent key={id} reagent={reagent} />
+                );
+              })
+            }
+          </Flex>
+        </Collapsible>
+      </Section>
+    </Stack.Item>
+  );
+};
+
+const Recipe = (props, context) => {
+  const { act, data } = useBackend(context);
+  const {
+    materials,
+    reagents,
+    efficiency,
+    stack_to_mineral,
+  } = data;
+  const {
+    recipe,
+  } = props;
+
+  const craft_amounts = [1, 5, 10, 50];
+  const substitutions = { "bluespace crystal": "bluespace_crystal" };
+
+  let max = recipe.max_amount;
+
+  const material_objects = Object.keys(recipe.materials).map(id => {
+    const material = materials[id] || materials[substitutions[id]];
+    const total = material.amount;
+    const amountNeeded = Math.floor(recipe.materials[id]
+      / (recipe.efficiency_affects ? efficiency : 1))
+      /stack_to_mineral;
+
+    const mat_max = Math.floor(total/amountNeeded);
+    max = Math.min(max, mat_max);
+    return (
+      <Box inline key={recipe.id+id} color={mat_max < 1 ? "red" : null}>
+        {amountNeeded} {material.name}
+      </Box>
+    );
+  });
+
+  const reagent_objects = Object.keys(recipe.reagents).map(id => {
+    const reagent = reagents[id];
+    const total = reagent?.volume || 0;
+    const recipeReagent = recipe.reagents[id];
+    const amountNeeded = Math.floor(recipeReagent.volume 
+      / (recipe.efficiency_affects ? efficiency : 1));
+    const mat_max = Math.floor(total/amountNeeded);
+    max = Math.min(max, mat_max);
+    return (
+      <Box inline key={recipe.id+id} color={mat_max < 1 ? "red" : null}>
+        {amountNeeded} {recipeReagent.name}
+      </Box>
+    );
+  });
+
+  const reducefn = (list, cur) => {
```

Comment:
This name clearly describes a function that adds items to a list with separators in between.

Reply:
Ordinarily it would just be a lambda, but I use two `reduce` calls so I separated it out. I guess I can make the name more descriptive though.

Reply:
it would've been much easier to just do `array.flat().join(' | ')`

Reply:
> it would've been much easier to just do `array.flat().join(' | ')`

I don't know the details of how JSX works, but since `join` turns it into a string, will there be no issues? I chose `reduce` to preserve the structure of the generated elements.

Reply:
Tested it, using `join` doesn't work, they just show up as `[object Object]`